packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -837,6 +853,273 @@ function rerenderReducer<S, I, A>(
+        source,
+        subscribe,
+      }: MutableSourceState<Source, Snapshot>),
+  );
+
+  const ref = useRefImpl({
+    getSnapshot: null,
+    source: null,
+    subscribe: null,
+  });
+
+  let didUnsubscribe = false;
+
+  // If we got a new source or subscribe function,
+  // we'll need to subscribe in a passive effect,
+  // and also check for any changes that fire between render and subscribe.
+  useEffectImpl(() => {
+    const shouldResubscribe =
+      subscribe !== ref.current.subscribe || source !== ref.current.source;
+
+    // We need to re-subscribe any time either the source or the subscribe function changes.
+    // We dont' need to re-subscribe when getSnapshot changes,
+    // but we will need the latest getSnapshot inside of our change handler-
+    // so we store a reference to it the ref as well.
+    ref.current = {
+      getSnapshot,
+      source,
+      subscribe,
+    };
+
+    if (shouldResubscribe) {
+      const handleChange = () => {
+        // It's possible that this callback will be invoked even after being unsubscribed,
+        // if it's removed as a result of a subscription event/update.
+        // In this case, React might log a DEV warning about an update from an unmounted component.
+        // We can avoid triggering that warning with this check.
+        if (didUnsubscribe) {
+          return;
+        }
+
+        const latestGetSnapshot = ((ref.current
+          .getSnapshot: any): MutableSourceGetSnapshotFn<Source, Snapshot>);
+        let newSnapshot = null;
+        let snapshotDidThrow = false;
+        try {
+          newSnapshot = latestGetSnapshot(source._source);
+        } catch (error) {
+          snapshotDidThrow = true;
+        }
+
+        setState(prevState => {
+          // Ignore values from stale sources!
+          // Since we subscribe an unsubscribe in a passive effect,
+          // it's possible that this callback will be invoked for a stale (previous) subscription.
+          // This check avoids scheduling an update for that stale subscription.
+          if (
+            prevState.source !== source ||
+            prevState.subscribe !== subscribe
+          ) {
+            return prevState;
+          }
+
+          // A selector might throw after a source mutation.
+          // e.g. it might try to read from a part of the store that no longer exists.
+          // In this case we should still schedule an update with React.
+          // Worst case the selector will throw again and then an error boundary will handle it.
+          if (snapshotDidThrow) {
+            return {...prevState};
+          }
+
+          // If the value hasn't changed, no update is needed.
+          // Return state as-is so React can bail out and avoid an unnecessary render.
+          if (prevState.snapshot === newSnapshot) {
+            return prevState;
+          }
+
+          setState({
+            ...prevState,
+            snapshot: newSnapshot,
+          });
```

Comment:
Should this be returning new state rather than calling setState in a setState callback? [I encounter an error in this CodeSandbox when incrementing the count twice.](https://codesandbox.io/s/sparkling-firefly-0r1xb)

Reply:
I encountered probably the same error: `TypeError: Cannot read property 'source' of undefined`

Reply:
Ah, this looks like maybe a rebase mistake of some sort...not sure what's going on here. Yes, it should just be a return. Thanks for pointing it out :)